controllers/vmsingle_controller.go

```diff
@@ -73,6 +75,44 @@ func (r *VMSingleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (r
 	if instance.Spec.ParsingError != "" {
 		return handleParsingError(instance.Spec.ParsingError, instance)
 	}
+	lastAppliedSingleSpec, err := instance.GetLastAppliedSpec()
+	if err != nil {
+		reqLogger.Error(err, "cannot parse last applied single spec")
+	}
+	singleChanges := deep.Equal(lastAppliedSingleSpec, &instance.Spec)
+	if len(singleChanges) == 0 {
```

Comment:
Status flow:
1) changes detected -> status `expanding`
2) CreateOrUpdateVMSingle returned an error -> status `failed` with reason.
3) CreateOrUpdateVMSingle and other functions returned no error -> status `operational` and update lastAppliedSpec.

Reply:
> I think, we shouldn't skip reconciliation for single. Since, `VMSingle` object could be updated without changing an annotation. By kubectl edit for instance.
> 
> Tracking changes is needed for changing object status.
> 
> E.g. if changes detected, update status for `expanding`.

kubectl edit instance will make the `len(singleChanges)>0`, doesn't it?
I think if there is no singleChanges, means the single.spec has no change. Since annotation only update when there is a change and reconcil is done in here.
https://github.com/VictoriaMetrics/operator/blob/2771c947528850e6b8cd83d493d5246a8e04abce/controllers/vmsingle_controller.go#L153-L160
So I only check the pod status to update the status of single, and return anyway to skip the rest of reconcil.

>CreateOrUpdateVMSingle and other functions returned no error -> status operational and update lastAppliedSpec.

I have different `operational` flow here, only mark single as  `operational` when readyReplicas == updatedReplicas == wantedReplicas.
Maybe it's too strict then.